mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -171,12 +192,23 @@ object KMeansModel extends MLReadable[KMeansModel] {
 
   /** [[MLWriter]] instance for [[KMeansModel]] */
   private[KMeansModel] class KMeansModelWriter(instance: KMeansModel) extends MLWriter {
+    import org.json4s.JsonDSL._
 
     private case class Data(clusterCenters: Array[Vector])
 
     override protected def saveImpl(path: String): Unit = {
-      // Save metadata and Params
-      DefaultParamsWriter.saveMetadata(instance, path, sc)
+      if (instance.isSet(instance.initialModel)) {
+        val initialModelPath = new Path(path, "initialModel").toString
+        val initialModel = instance.getInitialModel
+        initialModel.save(initialModelPath)
+
+        // Save metadata and Params
+        DefaultParamsWriter.saveMetadata(instance, path, sc, Some("hasInitialModel" -> true))
```

Comment:
Well, is it possible to handle this in `DefaultParamsWriter.saveMetadata`? The extra logic here seems to be boilerplate.